vumi/persist/txriak_manager.py

```diff
@@ -155,31 +159,60 @@ def set_user_metadata(self, usermeta):
         self._riak_obj.usermeta = usermeta
 
     def get_bucket(self):
-        return VumiTxRiakBucket(self._riak_obj.bucket)
+        return VumiTxRiakBucket(self._riak_obj.bucket, self._manager)
 
     # Methods that touch the network.
 
     def store(self):
-        d = deferToThread(self._riak_obj.store)
-        d.addCallback(type(self))
+        d = self._manager.deferToThread(self._riak_obj.store)
+        d.addCallback(type(self), self._manager)
         return d
 
     def reload(self):
-        d = deferToThread(self._riak_obj.reload)
-        d.addCallback(type(self))
+        d = self._manager.deferToThread(self._riak_obj.reload)
+        d.addCallback(type(self), self._manager)
         return d
 
     def delete(self):
-        d = deferToThread(self._riak_obj.delete)
-        d.addCallback(type(self))
+        d = self._manager.deferToThread(self._riak_obj.delete)
+        d.addCallback(type(self), self._manager)
         return d
 
 
+def _daemon_threadpool(*args, **kw):
+    """
+    Create a threadpool that sets its threads to daemon mode so they don't
+    block process exit if they aren't properly cleaned up.
```

Comment:
This has nothing to do with whether there's an incomplete Riak operation during shutdown, it's about what happens to threads that are not explicitly cleaned up.

If you fail to clean up a manager with non-daemon threads, you're guaranteed to have the process hang indefinitely (unless you hit it with SIGKILL) because the still-running threadpool will have threads that are waiting for more work to do. An in-progress Riak operation could still complete after the main thread terminates, but it would have nowhere to send its result and its thread will still hang around afterwards waiting for more work that will never arrive.

If you fail to clean up a manager with daemon threads, there's a risk of killing an in-progress Riak operation that you're not waiting for during service shutdown. This is exactly the same as a completely async operation where you're not waiting for things to finish during service shutdown.